### PR TITLE
Bug 26140: Introducing param to skip attribute refresh when valid

### PR DESCRIPTION
**Pull request by vardhanh123**
unitTested

10.10.103.38

NFS Ganesh

### CODE REVIEW

File: src/cache_inode/cache_inode_rdwr.c [hunk 1]
```diff
@@ -268,8 +268,14 @@ cache_inode_status_t cache_inode_rdwr(cache_entry_t *entry,
...
 	PTHREAD_RWLOCK_wrlock(&entry->attr_lock);
 	attributes_locked = true;
-	if (io_direction == CACHE_INODE_WRITE ||
-	    io_direction == CACHE_INODE_WRITE_PLUS) {
+	bool checkAttrs = nfs_param.nfsv4_param.check_valid_attrs ?
+				cache_inode_is_attrs_valid(entry) : false;
+	LogDebug(COMPONENT_CACHE_INODE,
+			"cache_inode_rdwr: check_valid_attrs=%d, Attribute valid=%d",
+			nfs_param.nfsv4_param.check_valid_attrs,
+			cache_inode_is_attrs_valid(entry));
```
Comment (SriKrishnachowdaryk) [inline src/cache_inode/cache_inode_rdwr.c]: Do not call this again. print CheckAttrs and check_valid_attrs instead.

**Review by SriKrishnachowdaryk — CHANGES_REQUESTED** [review]
Commit message should describe more. pls explain why we did what we did in this change.

Done

File: src/cache_inode/cache_inode_rdwr.c [hunk 2]
```diff
@@ -268,8 +268,14 @@ cache_inode_status_t cache_inode_rdwr(cache_entry_t *entry,
...
-	if (io_direction == CACHE_INODE_WRITE ||
-	    io_direction == CACHE_INODE_WRITE_PLUS) {
+	bool checkAttrs = nfs_param.nfsv4_param.check_valid_attrs ?
+				cache_inode_is_attrs_valid(entry) : false;
+	LogDebug(COMPONENT_CACHE_INODE,
+			"cache_inode_rdwr: check_valid_attrs=%d, Valid Attributes =%d",
+			nfs_param.nfsv4_param.check_valid_attrs,
+			checkAttrs);
+	if ((io_direction == CACHE_INODE_WRITE ||
+	    io_direction == CACHE_INODE_WRITE_PLUS) && !checkAttrs) {
```
Comment (satgs) [inline src/cache_inode/cache_inode_rdwr.c]: We are modifying atime if checkattrs = true. That's undesirable.

File: src/cache_inode/cache_inode_rdwr.c [hunk 3]
```diff
@@ -270,7 +270,16 @@ cache_inode_status_t cache_inode_rdwr(cache_entry_t *entry,
...
-		status = cache_inode_refresh_attrs(entry);
+
+		bool canRefresh = nfs_param.nfsv4_param.post_write_attrs_update ?
+			true : !cache_inode_is_attrs_valid(entry);
+		LogDebug(COMPONENT_CACHE_INODE,
+			"cache_inode_rdwr: post_write_attrs_update=%d, Can refresh attributes=%d",
+			nfs_param.nfsv4_param.post_write_attrs_update,
+			canRefresh);
+
+    if (canRefresh)
```
Comment (SriKrishnachowdaryk) [inline src/cache_inode/cache_inode_rdwr.c]: tabs

**Review by SriKrishnachowdaryk — CHANGES_REQUESTED** [review]

File: src/cache_inode/cache_inode_rdwr.c [hunk 4]
```diff
@@ -270,7 +270,16 @@ cache_inode_status_t cache_inode_rdwr(cache_entry_t *entry,
...
+
+		bool canRefresh = nfs_param.nfsv4_param.post_write_attrs_update ?
+			true : !cache_inode_is_attrs_valid(entry);
+		LogDebug(COMPONENT_CACHE_INODE,
+			"cache_inode_rdwr: post_write_attrs_update=%d, Can refresh attributes=%d",
+			nfs_param.nfsv4_param.post_write_attrs_update,
+			canRefresh);
+
+		if (canRefresh)
+		  status = cache_inode_refresh_attrs(entry);
```
Comment (SriKrishnachowdaryk) [inline src/cache_inode/cache_inode_rdwr.c]: tabs here.

**Review by SriKrishnachowdaryk — CHANGES_REQUESTED** [review]

**Review by SriKrishnachowdaryk — COMMENTED** [review]
LGTM

File: src/support/nfs_read_conf.c [hunk 5]
```diff
@@ -252,6 +252,8 @@ static struct config_item version4_params[] = {
 	CONF_ITEM_UI64("Dirent_Cache_Threshold", 0, UINT64_MAX,
 			DIRENT_CACHE_THRESHOLD_DEFAULT,
 			nfs_version4_parameter, dirent_cache_threshold),
+	CONF_ITEM_BOOL("Post_Write_Attrs_Update", true,
```
Comment (satgs) [inline src/support/nfs_read_conf.c]: Default value should be false.

File: src/cache_inode/cache_inode_rdwr.c [hunk 6]
```diff
@@ -270,7 +270,16 @@ cache_inode_status_t cache_inode_rdwr(cache_entry_t *entry,
...
-		status = cache_inode_refresh_attrs(entry);
+
+		bool canRefresh = nfs_param.nfsv4_param.post_write_attrs_update ?
+			true : !cache_inode_is_attrs_valid(entry);
+		LogDebug(COMPONENT_CACHE_INODE,
+			"cache_inode_rdwr: post_write_attrs_update=%d, Can refresh attributes=%d",
+			nfs_param.nfsv4_param.post_write_attrs_update,
+			canRefresh);
+
+		if (canRefresh)
```
Comment (satgs) [inline src/cache_inode/cache_inode_rdwr.c:281]: if (canRefresh) {
  status = refresh_attrs;
  if (status != SUCCESS) {
    goto out;
  }
}
/* else {
  status = SUCCESS;
} */

**Review by satgs — COMMENTED** [review]
please fix the changes and check it in.

**Pull request merged**